File: frontend/src/app/services/responsibility-centre.service.ts
/*
 * myRC - Responsibility Centre Service
 * Copyright (c) 2026 myRC Team
 * Licensed under MIT License
 */

import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse } from '@angular/common/http';
import { BehaviorSubject, Observable, Subject, throwError } from 'rxjs';
import { catchError } from 'rxjs/operators';
import { ResponsibilityCentreDTO } from '../models/responsibility-centre.model';

@Injectable({
  providedIn: 'root'
})
export class ResponsibilityCentreService {
  private apiUrl = '/api/responsibility-centres';
  private selectedRCSubject = new BehaviorSubject<number | null>(
    this.getStoredSelectedRC()
  );
  public selectedRC$ = this.selectedRCSubject.asObservable();

  private selectedFYSubject = new BehaviorSubject<number | null>(
    this.getStoredSelectedFY()
  );
  public selectedFY$ = this.selectedFYSubject.asObservable();

  /** Emits when RC data is updated (e.g. feature toggles changed), so subscribers can refresh. */
  private rcUpdatedSubject = new Subject<number>();
  public rcUpdated$ = this.rcUpdatedSubject.asObservable();

  /** Notify subscribers that an RC has been updated (e.g. after toggling features). */
  notifyRCUpdated(rcId: number): void {
    this.rcUpdatedSubject.next(rcId);
  }

  constructor(private http: HttpClient) {}

  getAllResponsibilityCentres(): Observable<ResponsibilityCentreDTO[]> {
    return this.http.get<ResponsibilityCentreDTO[]>(this.apiUrl, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  getResponsibilityCentre(id: number): Observable<ResponsibilityCentreDTO> {
    return this.http.get<ResponsibilityCentreDTO>(`${this.apiUrl}/${id}`, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  createResponsibilityCentre(
    name: string,
    description: string
  ): Observable<ResponsibilityCentreDTO> {
    return this.http.post<ResponsibilityCentreDTO>(this.apiUrl, {
      name,
      description
    }, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  updateResponsibilityCentre(
    id: number,
    name: string,
    description: string
  ): Observable<ResponsibilityCentreDTO> {
    return this.http.put<ResponsibilityCentreDTO>(`${this.apiUrl}/${id}`, {
      name,
      description
    }, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  deleteResponsibilityCentre(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  cloneResponsibilityCentre(id: number, newName: string): Observable<ResponsibilityCentreDTO> {
    return this.http.post<ResponsibilityCentreDTO>(`${this.apiUrl}/${id}/clone`, {
      newName
    }, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  grantAccess(
    rcId: number,
    username: string,
    accessLevel: string
  ): Observable<any> {
    return this.http.post(`${this.apiUrl}/${rcId}/access/grant`, {
      username,
      accessLevel
    }, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  revokeAccess(rcId: number, username: string): Observable<any> {
    return this.http.post(`${this.apiUrl}/${rcId}/access/revoke`, {
      username
    }, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  getResponsibilityCentreAccess(rcId: number): Observable<any[]> {
    return this.http.get<any[]>(`${this.apiUrl}/${rcId}/access`, { withCredentials: true })
      .pipe(catchError(this.handleError));
  }

  setSelectedRC(rcId: number): void {
    localStorage.setItem('selectedRC', rcId.toString());
    this.selectedRCSubject.next(rcId);
  }

  getSelectedRC(): number | null {
    return this.selectedRCSubject.value;
  }

  setSelectedFY(fyId: number): void {
    localStorage.setItem('selectedFY', fyId.toString());
    this.selectedFYSubject.next(fyId);
  }

  getSelectedFY(): number | null {
    return this.selectedFYSubject.value;
  }

  clearSelection(): void {
    localStorage.removeItem('selectedRC');
    localStorage.removeItem('selectedFY');
    this.selectedRCSubject.next(null);
    this.selectedFYSubject.next(null);
  }

  setTrainingEnabled(rcId: number, enabled: boolean): Observable<ResponsibilityCentreDTO> {
    return this.http.patch<ResponsibilityCentreDTO>(
      `${this.apiUrl}/${rcId}/training-enabled`,
      { enabled },
      { withCredentials: true }
    ).pipe(catchError(this.handleError));
  }

  setTravelEnabled(rcId: number, enabled: boolean): Observable<ResponsibilityCentreDTO> {
    return this.http.patch<ResponsibilityCentreDTO>(
      `${this.apiUrl}/${rcId}/travel-enabled`,
      { enabled },
      { withCredentials: true }
    ).pipe(catchError(this.handleError));
  }

  setTrainingIncludeInSummary(rcId: number, include: boolean): Observable<ResponsibilityCentreDTO> {
    return this.http.patch<ResponsibilityCentreDTO>(
      `${this.apiUrl}/${rcId}/training-include-in-summary`,
      { include },
      { withCredentials: true }
    ).pipe(catchError(this.handleError));
  }

  setTravelIncludeInSummary(rcId: number, include: boolean): Observable<ResponsibilityCentreDTO> {
    return this.http.patch<ResponsibilityCentreDTO>(
      `${this.apiUrl}/${rcId}/travel-include-in-summary`,
      { include },
      { withCredentials: true }
    ).pipe(catchError(this.handleError));
  }

  private getStoredSelectedRC(): number | null {
    const stored = localStorage.getItem('selectedRC');
    return stored ? parseInt(stored, 10) : null;
  }

  private getStoredSelectedFY(): number | null {
    const stored = localStorage.getItem('selectedFY');
    return stored ? parseInt(stored, 10) : null;
  }

  /**
   * Handle HTTP errors.
   *
   * @param error The HTTP error response
   * @returns Observable that throws an error
   */
  private handleError(error: HttpErrorResponse): Observable<never> {
    let errorMessage = 'An unknown error occurred';

    if (error.error instanceof ErrorEvent) {
      // Client-side error
      errorMessage = `Client error: ${error.error.message}`;
    } else {
      // Server-side error - check for specific error messages from backend
      const backendMessage = error.error?.message || error.error?.error || '';
      
      switch (error.status) {
        case 400:
          // Check if it's a duplicate name error
          if (backendMessage.toLowerCase().includes('already exists') || 
              backendMessage.toLowerCase().includes('duplicate')) {
            errorMessage = backendMessage;
          } else {
            errorMessage = backendMessage || 'Invalid request. Please check your input.';
          }
          break;
        case 401:
          errorMessage = 'Your session has expired. Please log in again.';
          break;
        case 403:
          errorMessage = 'You do not have permission to perform this action.';
          break;
        case 404:
          errorMessage = 'Responsibility Centre not found.';
          break;
        case 409:
          // Conflict - typically duplicate name
          errorMessage = backendMessage || 'A Responsibility Centre with this name already exists.';
          break;
        case 500:
          errorMessage = 'Server error occurred. Please try again later.';
          break;
        default:
          errorMessage = backendMessage || `Server error: ${error.status} - ${error.statusText}`;
      }
    }

    console.error('RC Service Error:', errorMessage);
    return throwError(() => new Error(errorMessage));
  }
}

File: backend/src/main/java/com/myrc/service/ResponsibilityCentreService.java
/*
 * myRC - Responsibility Centre Service Interface
 * Copyright (c) 2026 myRC Team
 * Licensed under MIT License
 */
package com.myrc.service;

import com.myrc.dto.ResponsibilityCentreDTO;
import com.myrc.model.RCAccess;
import java.util.List;
import java.util.Optional;

/**
 * Service interface for managing Responsibility Centres.
 * Defines operations for creating, reading, updating, and managing access to responsibility centres.
 *
 * @author myRC Team
 * @version 1.0.0
 * @since 2026-01-17
 */
public interface ResponsibilityCentreService {

  /**
   * Get all responsibility centres accessible to a user.
   * Includes RCs owned by the user, RCs with direct user access, and RCs
   * accessible via group membership (LDAP group DNs).
   *
   * @param username the username of the user
   * @param groupIdentifiers list of LDAP group DNs the user belongs to (may be empty)
   * @return list of responsibility centres the user owns or has access to
   */
  List<ResponsibilityCentreDTO> getUserResponsibilityCentres(String username,
      List<String> groupIdentifiers);

  /**
   * Create a new responsibility centre.
   *
   * @param username the username of the owner
   * @param name the name of the responsibility centre
   * @param description the description of the responsibility centre
   * @return the created responsibility centre DTO
   */
  ResponsibilityCentreDTO createResponsibilityCentre(String username, String name, 
      String description);

  /**
   * Get a specific responsibility centre by ID.
   * Checks ownership, direct user access, and group-based access.
   *
   * @param rcId the ID of the responsibility centre
   * @param username the username of the requesting user
   * @param groupIdentifiers list of LDAP group DNs the user belongs to (may be empty)
   * @return optional containing the responsibility centre if found and user has access
   */
  Optional<ResponsibilityCentreDTO> getResponsibilityCentre(Long rcId, String username,
      List<String> groupIdentifiers);

  /**
   * Update a responsibility centre.
   *
   * @param rcId the ID of the responsibility centre
   * @param username the username of the requesting user (must be owner)
   * @param name the new name
   * @param description the new description
   * @return optional containing the updated responsibility centre if successful
   */
  Optional<ResponsibilityCentreDTO> updateResponsibilityCentre(Long rcId, String username,
      String name, String description);

  /**
   * Delete a responsibility centre.
   *
   * @param rcId the ID of the responsibility centre
   * @param username the username of the requesting user (must be owner)
   * @return true if deletion was successful, false otherwise
   */
  boolean deleteResponsibilityCentre(Long rcId, String username);

  /**
   * Grant access to a responsibility centre to another user.
   *
   * @param rcId the ID of the responsibility centre
   * @param ownerUsername the username of the owner
   * @param grantedToUsername the username of the user to grant access to
   * @param accessLevel the access level (READ_ONLY or READ_WRITE)
   * @return optional containing the created access record if successful
   */
  Optional<RCAccess> grantAccess(Long rcId, String ownerUsername, String grantedToUsername,
      String accessLevel);

  /**
   * Revoke access to a responsibility centre from a user.
   *
   * @param rcId the ID of the responsibility centre
   * @param ownerUsername the username of the owner
   * @param revokeFromUsername the username of the user to revoke access from
   * @return true if revocation was successful, false otherwise
   */
  boolean revokeAccess(Long rcId, String ownerUsername, String revokeFromUsername);

  /**
   * Get all access records for a responsibility centre.
   *
   * @param rcId the ID of the responsibility centre
   * @param ownerUsername the username of the owner
   * @return list of access records for the responsibility centre
   */
  List<RCAccess> getResponsibilityCentreAccess(Long rcId, String ownerUsername);

  /**
   * Clone a responsibility centre with a new name.
   * Creates a new RC owned by the requesting user with the same structure but no fiscal years.
   *
   * @param sourceRcId the ID of the responsibility centre to clone
   * @param username the username of the user creating the clone
   * @param newName the name for the cloned responsibility centre
   * @return the cloned responsibility centre DTO
   */
  ResponsibilityCentreDTO cloneResponsibilityCentre(Long sourceRcId, String username, String newName);

  /**
   * Toggle the training enabled status for a responsibility centre.
   *
   * @param rcId the ID of the responsibility centre
   * @param username the username of the requesting user (must be owner)
   * @param enabled the new training enabled status
   * @return optional containing the updated responsibility centre if successful
   */
  Optional<ResponsibilityCentreDTO> setTrainingEnabled(Long rcId, String username, boolean enabled);

  /**
   * Toggle the travel enabled status for a responsibility centre.
   *
   * @param rcId the ID of the responsibility centre
   * @param username the username of the requesting user (must be owner)
   * @param enabled the new travel enabled status
   * @return optional containing the updated responsibility centre if successful
   */
  Optional<ResponsibilityCentreDTO> setTravelEnabled(Long rcId, String username, boolean enabled);

  /**
   * Toggle training include in summary for a responsibility centre.
   */
  Optional<ResponsibilityCentreDTO> setTrainingIncludeInSummary(Long rcId, String username, boolean include);

  /**
   * Toggle travel include in summary for a responsibility centre.
   */
  Optional<ResponsibilityCentreDTO> setTravelIncludeInSummary(Long rcId, String username, boolean include);
}
